CenterForOpenScience/ember-osf-web
# [ENG-458][ENG-437] View-only links

### NOTE:
- Replaces #660 
- Not testable without BE changes (but could be safely merged/released without them) https://github.com/CenterForOpenScience/osf.io/pull/9037


## Purpose
Support view-only links.


## Summary of Changes
- Add `view_only` query param to application route
- Store VOL info on the `current-user` service
- Include `view_only` in all API requests, all links within OSF
- Display a banner with a button to stop using a VOL
- If the VOL is anonymized, display "Anonymous contributors" in place of any contributor list
- Anonymize user-contributed fields in license text
- Anonymize response to "Authors" question in registration form


## Side Effects
Added some VOL logic to the `@checkAuth` decorator/mixin because the VOL token should be set *before* the mixed-in `beforeModel` to correctly handle all auth flows. This violates `@checkAuth`'s encapsulation, but the need for this seems to indicate it shouldn't be a decorator/mixin at all. In the future, we should refactor it to a service method called in the application route's `beforeModel`


## Feature Flags
`ember_registries_detail_page` for testing on registrations


## QA Notes
Backend branch required for testing VOLs on the new registries overview page: https://github.com/CenterForOpenScience/osf.io/pull/9020

Registries scenarios:
- Non-anonymized VOL: anyone can view (even embargoed reg, even logged out)
- Anonymized VOL: anyone can view (even embargoed reg, even logged out), but no contributor info available -- contributor lists say "Anonymous contributors"

Other things:
- VOL banner:
  - Because VOLs in the API are more comprehensive than in legacy, we always display a VOL banner on embosf'd pages.
  - Legacy pages don't have a matching banner. This is ok.
  - Clicking "Leave this view" in the banner removes the `view_only` param from the URL and takes you to the landing page (dashboard, if logged in).
- When you start with a VOL and browse around OSF (without clicking "Leave this view") `view_only=blahblahblah` should stay in the URL
- Anonymized VOLs hide *all* contributor info. In particular:
  - Licenses with user-contributor fields will anonymize those fields
  - Questions titled "Authors" on the registration form are anonymized (only Prereg Challenge and Registered Report Protocol Preregistration are currently affected)
  - If you start with an anonymized VOL and navigate e.g. to the dashboard, all projects in your dashboard list will be anonymized too.




## Ticket


Feature: https://openscience.atlassian.net/browse/ENG-458
Subtask: https://openscience.atlassian.net/browse/ENG-437

# Reviewer Checklist
